grid2op/multi_agent/multiAgentEnv.py

```diff
@@ -0,0 +1,562 @@
+                actionClass=self._cent_env._actionClass,
+                legal_action=self._cent_env._game_rules.legal_action,
+            )
+    
+    
+    def _update_observations(self, _update_state = True):
+        """Update agents' observations from the global observation given by the self._cent_env
+
+        Args:
+            observation (BaseObservation): _description_
+        """
+        #for agent in self.agents:
+        #    self.observations[agent] = observation.copy()
+        
+        if self._cent_env.__closed:
+            raise EnvError("This environment is closed. You cannot use it anymore.")
+        if not self.__is_init:
+            raise EnvError(
+                "This environment is not initialized. You cannot retrieve its observation."
+            )
+        
+        for agent in self.agents:
+            self.observations[agent] = self.observation_spaces[agent](
+                env=self._cent_env, _update_state=_update_state
+            )
+    
+    def _verify_domains(self, domains : MADict) :
+        """It verifies if substation ids are valid
+
+        Args:
+            domains (dict): 
+                - key : agents' names
+                - value : list of substation ids convertible to int (allowed types are list, set and np.ndarray)
+                These domains must be non empty and constitute a partition, i.e., the union is the set of all substation ids 
+                in the global env and any intersection between two domains is empty.
+        """
+        sum_subs = 0
+        for agent in domains.keys():
+            if not (isinstance(domains[agent], list)
+                    or isinstance(domains[agent], set)
+                    or isinstance(domains[agent], np.ndarray)
+            ):
+                raise DomainException(f"Agent id {agent} : The domain must be a list or a set of substation indices")
+            
+            if len(domains[agent]) == 0:
+                raise DomainException(f"Agent id {agent} : The domain is empty !")
+            
+            for i in range(len(domains[agent])) : 
+                try:
+                    domains[agent][i] = int(domains[agent][i])
+                except Exception as e:
+                    raise DomainException(f"Agent id {agent} : The id must be of type int or convertible to an int. Type {type(domains[agent][i])} is not valid")
+                if domains[agent][i] < 0 or domains[agent][i] > self._cent_env.n_sub:
+                    raise DomainException(f"Agent id {agent} : The substation's id must be between 0 and {len(self._cent_env.name_sub)-1}, but {domains[agent][i]} has been given")
+            sum_subs += len(domains[agent])
+
+        if sum_subs != self._cent_env.n_sub:
+            raise DomainException(f"The sum of sub id lists' length must be equal to _cent_env.n_sub = {self._cent_env.n_sub} but is {sum_subs}")
+    
+    
+    def observation_space(self, agent : AgentID):
```

Comment:
Don't "half do" things, in this case either you typint (which is a good thing !) or you dont.

So if you give input parameters types, give also the output parameters type.
If you don't, put a "todo typing" somewhere to remember to put them